web/packages/design/src/utils/useResizeObserver.ts

```diff
@@ -16,46 +16,78 @@
+import {
+  useCallback,
+  useEffect,
+  useMemo,
+  useRef,
+  type RefCallback,
+} from 'react';
 
 /**
- * useResizeObserver sets up a ResizeObserver for ref and calls callback on each resize.
+ * useResizeObserver creates a ResizeObserver which fires a callback
+ * when the provided element is resized. The callback is called with the
+ * ResizeObserverEntry. The observer is disconnected when the element is
+ * unmounted.
  *
- * It does not fire if ref.current.contentRect.height is zero, to account for a special case in
- * Connect where tabs are hidden using `display: none;`.
+ * Returns a ref callback to attach to the element to be observed – the element
+ * must be a HTMLElement but may be conditionally null.
  *
- * Uses a layout effect underneath. If ref is conditionally rendered, set enabled to false when ref
- * is null.
+ * @example
+ * // Basic usage
+ * const Component = () => {
+ *   const handleResize = (entry: ResizeObserverEntry) => {
+ *     console.log({ entryContentRect: entry.contentRect });
+ *   };
+ *
+ *   const ref = useResizeObserver<HTMLDivElement>(handleResize);
+ *
+ *   return <div ref={ref}>This div is being observed</div>;
+ * };
  */
-export function useResizeObserver(
-  ref: RefObject<HTMLElement>,
+export const useResizeObserver = <T extends HTMLElement = HTMLElement>(
   callback: (entry: ResizeObserverEntry) => void,
-  { enabled = true }
-): void {
-  const effect = useCallback(() => {
-    if (!ref.current || !enabled) {
-      return;
-    }
+  { fireOnZeroHeight = true } = {}
+): RefCallback<T> => {
+  const callbackRef = useRef(callback);
 
-    const observer = new ResizeObserver(entries => {
-      const entry = entries[0];
+  useEffect(() => {
+    callbackRef.current = callback;
+  }, [callback]);
 
-      // In Connect, when a tab becomes active, its outermost DOM element switches from `display:
-      // none` to `display: flex`. This callback is then fired with the height reported as zero.
-      // To avoid unnecessary calls to callback, return early here.
-      if (entry.contentRect.height === 0) {
-        return;
-      }
+  const observer = useMemo(
+    () =>
+      new ResizeObserver(([entry]) => {
+        if (!entry || (!fireOnZeroHeight && entry.contentRect.height === 0))
+          return;
+        callbackRef.current?.(entry);
+      }),
+    [fireOnZeroHeight]
+  );
 
-      callback(entry);
-    });
+  const elementRef = useRef<T | null>(null);
 
-    observer.observe(ref.current);
+  useEffect(() => {
+    const element = elementRef.current;
+    if (element) {
+      observer.observe(element);
+    }
 
-    return () => {
-      observer.disconnect();
-    };
-  }, [callback, ref, enabled]);
+    return () => observer.disconnect();
+  }, [observer]);
 
-  useLayoutEffect(effect, [effect]);
-}
+  return useCallback(
+    (element: T | null) => {
+      if (elementRef.current && !elementRef.current.isEqualNode(element)) {
```

Comment:
Do we need to compare `elementRef.current` to `element`? Could we instead simplify the body of the ref callback to something like this?

```tsx
      if (element) {
        observer.observe(element);
        elementRef.current = element;
      } else {
        elementRef.current && observer.unobserve(elementRef.current);
        elementRef.current = null;
      }
```

This mirrors how ref callbacks work in v18 and we don't have to pay the price of running `Node.prototype.isEqualNode`.

Reply:
Having a simple ref callback body like this doesn't cause unnecessary calls to observe/unobserve, e.g. if the component using `useResizeObserver` rerenders but the observed element doesn't change, the ref callback doesn't get called.

One thing is the "price" of calling `isEqualNode', which probably isn't that high, but the other thing is deciding what does it mean for nodes to be equal within the React world. Instead of having a say in that ourselves, we can let React decide – if the underlying node is different, it's going to call the ref callback with null and then again with the new node.

Reply:
you're right, we don't need to – went ahead and simplified the ref cb, and removed that redundant observation in the second effect.